test/unit/BPool.t.sol

```diff
@@ -1350,31 +1350,156 @@ contract BPool_Unit_JoinswapPoolAmountOut is BasePoolTest {
     bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, _maxAmountIn);
   }
 
-  function test_Revert_NotFinalized() private view {}
+  function test_Revert_NotFinalized() public {
+    _setFinalize(false);
 
-  function test_Revert_NotBound() private view {}
+    vm.expectRevert('ERR_NOT_FINALIZED');
+    bPool.joinswapPoolAmountOut(tokenIn, 0, 0);
+  }
 
-  function test_Revert_MaxApprox() private view {}
+  function test_Revert_NotBound(
+    JoinswapPoolAmountOut_FuzzScenario memory _fuzz,
+    address _tokenIn
+  ) public happyPath(_fuzz) {
+    vm.assume(_tokenIn != VM_ADDRESS);
 
-  function test_Revert_LimitIn() private view {}
+    vm.expectRevert('ERR_NOT_BOUND');
+    bPool.joinswapPoolAmountOut(_tokenIn, 0, 0);
+  }
 
-  function test_Revert_MaxInRatio() private view {}
+  function test_Revert_MathApprox(JoinswapPoolAmountOut_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    _fuzz.poolAmountOut = 0;
 
-  function test_Revert_Reentrancy() private view {}
+    vm.expectRevert('ERR_MATH_APPROX');
+    bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, type(uint256).max);
+  }
 
-  function test_Set_Balance() private view {}
+  function test_Revert_LimitIn(JoinswapPoolAmountOut_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    uint256 _tokenAmountIn = calcSingleInGivenPoolOut(
+      _fuzz.tokenInBalance,
+      _fuzz.tokenInDenorm,
+      _fuzz.totalSupply,
+      _fuzz.totalWeight,
+      _fuzz.poolAmountOut,
+      _fuzz.swapFee
+    );
 
-  function test_Emit_LogJoin() private view {}
+    vm.expectRevert('ERR_LIMIT_IN');
+    bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, _tokenAmountIn - 1);
+  }
 
-  function test_Mint_PoolShare() private view {}
+  function test_Revert_MaxInRatio(JoinswapPoolAmountOut_FuzzScenario memory _fuzz) public {
+    // TODO: improve and remove hardcoded values
+    _fuzz.poolAmountOut = 39_000_000 ether;
+    _fuzz.tokenInBalance = 110 ether;
+    _fuzz.tokenInDenorm = MIN_WEIGHT;
+    _fuzz.totalSupply = 38_000_000_000 ether;
+    _fuzz.totalWeight = 392 ether;
+    _fuzz.swapFee = 0.03 ether;
 
-  function test_Push_PoolShare() private view {}
+    _setValues(_fuzz);
```

Comment:
ok, this is tough, because we need to calculate `tokenAmountIn` from `balanceIn`, and then ask `balanceIn` to be below a certain threshold related to `tokenAmountIn`, i'd suggest the following:
- assumeHappyPath with fuzzed variables and calculate `tokenAmountIn`
- add a fuzzed `uint256` that is bounded to some value related to `tokenAmountIn` and the current fuzzed `balanceIn`
- `_fuzz.tokenInBalance = _tokenInBalance`
- `setValues(_fuzz)`

but gotta agree that perhaps this isn't possible always, and there's only a range of values in which this error can exist, interesting that you've found a hardcoded combination, how did you achieve that?

Reply:
I managed to make it fail once and copied the fuzzed values. Then adjusted those values to make it look more nice, rounding up or down.